### PR TITLE
Add Haskell example

### PR DESCRIPTION
**Pull request by ElyKar**

### CODE REVIEW

File: src/haskell/README.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+
+## Haskell
+
+Haskell is a strong, static, nominal and strucural typing language. The trick is to define a Duck typeclass, then make Mallard an instance of it. Wolf could very well be an instance of Duck, however it makes more sense to use a DuckCosplay that allows hime to quake and dance like a Duck.
```
Comment (arnobl) [inline src/haskell/README.md]: hime -> him

File: src/haskell/duck.hs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+class Duck a where
+    quake :: a -> String
+    dance :: a -> String
+
+data Wolf = Wolf
+
+eat :: (Duck t) => Wolf -> t -> String
+eat _ _ = " 😈 "
+
+{- In order to become a duck, Wolf has to use the Adapter pattern. In
```
Comment (arnobl) [inline src/haskell/duck.hs:11]: As far as I understand, Haskell has a structural typing system (see e.g. http://tratt.net/laurie/research/pubs/html/tratt__dynamically_typed_languages/). I cannot, however, find any example about that. Any idea?

**Review by arnobl — CHANGES_REQUESTED** [review]
Thx. Minor changes to apply before a merge. I cannot figure out whether Haskell has a structural typing (I cannot find examples).

File: src/haskell/duck.hs (hunk 2 above)
Comment (ElyKar) [reply src/haskell/duck.hs:11]: I've seen about an equal number of people declaring that Haskell has or has no structural typing. The best answer I found is [here](http://stackoverflow.com/questions/21071603/why-does-haskell-not-have-records-with-structural-typing). It seems accepted that Haskell can provide a form of structural typing through the use of typeclasses : Mallard and DuckCosplay types can be considered as mere Duck because they instantiate it.
Apart from that, the compiler is very careful about types, which makes Haskell typing system sometimes cruelly nominal (e.g. won't compile when doing (1 :: Int) + (1 :: Integer) ).

Comment (arnobl) [reply src/haskell/duck.hs:11]: Indeed, because Haskell does not apply coercion by default. For example with ocaml, I needed to explicitly call the coercion operator to check and convert a wolf as a duck (thewolf :> duck). Maybe something like that exists in Haskell but I cannot find how to do that.
Can you check whether https://wiki.haskell.org/GHC/Coercible can solve the problem?

Cheers
(but do not forget to work your exams...)

**Review by arnobl — APPROVED** [review]

File: src/haskell/duck.hs (hunk 2 above)
Comment (ElyKar) [reply src/haskell/duck.hs:11]: After checking a bit about coercion in Haskell, I don't think it can be applied to this use case (or at least, not in a straight-forward way). Duck would have to be a type (impossible to coerce from a type (Wolf) to a type constraint (Duck)). And even then, it would mean making usual functions with pattern matching, and we would lose the information we have about Wolf being a Wolf.
A (untested) workaround would be to make Wolf, Duck and Mallard records, but I don't know if coercion would still be available, and if it would solve the problem adequatly. Plus, this would be ugly.

TL; DR : I don't think coercion is applicable in this case. (Though I'm far from mastering every aspect of Haskell type system).

**Comment by arnobl** [comment]
Thx. If structural typing can be applied here, maybe an Haskeller would submit a PR.
